---
title: "bulk ATAC"
output: html_document
author: "Pierre-Luc Germain"
date: "2024-10-22"
---

```{r setup}
knitr::opts_chunk$set(dev="CairoPNG")
suppressPackageStartupMessages({
  library(sechm)
  library(epiwraps)
  library(ggplot2)
  library(cowplot)
  library(patchwork)
  library(ensembldb)
  library(AnnotationHub)
  library(cowplot)
})
ah <- AnnotationHub(cache="/common/AnnotationHub/", localHub=TRUE)
ensdb <- ah[["AH89211"]]

source("../functions/functions.R")
source("../functions/clusterProfiling.R")
TFcols <- c("FALSE"="darkgrey", "TRUE"="darkblue")
```

```{r}
rse <- readRDS("../bulkATAC/SE.proc.rds")
rse$Sex <- factor(rse$Sex, c("F","M"), c("female","male"))
rse$Treatment <- factor(rse$Treatment, unique(rse$Treatment), c("Handling+ARS","CRS+ARS"))
rse$TimePoint <- rse$Group
levels(rse$TimePoint) <- c("0min","45min")
rda <- getDEA(rse, "stress")
rda <- cbind(rda, as.data.frame(rowData(rse)[row.names(rda),12:17]))

mds <- readRDS("../bulkATAC/mds_excludingSex.rds")
mds$TimePoint <- mds$Group
mds$Treatment <- factor(mds$Treatment, unique(mds$Treatment), levels(rse$Treatment))
levels(mds$TimePoint) <- levels(rse$TimePoint)
crscols <- c("Handling+ARS" = "darkgrey", "CRS+ARS" = "#7964ffff")
mds$Group <- gsub("+","+\n",paste(mds$Treatment,mds$TimePoint),fixed=TRUE)
mds <- mds[order(mds$TimePoint, mds$Treatment),]
mds$Group <- factor(mds$Group, unique(mds$Group))
cols <- setNames(c("darkgrey","lightblue","red","#7964ffff"), levels(mds$Group))
ve <- attr(mds,"varExplained")
mdp2 <- ggplot(mds, aes(x,y, colour=Group)) + 
  geom_hline(yintercept=-0.025, linetype="dashed", colour="darkgrey") + geom_point(size=3) +
    labs(x=paste0("MDS1 (",round(100*ve[1]),"% of variance)"),
       y=paste0("MDS2 (",round(100*ve[2]),"% of variance)")) + mytheme() +
  scale_colour_manual(values=cols)
mdp2.vl <- ggplot(mds, aes(TimePoint, y, fill=Group)) + geom_violin() + 
  stat_summary(position=position_dodge2(width=0.9)) + mytheme() +
  labs(x="TimePoint", y="MDS2") + scale_fill_manual(values=cols)
mdp <- mdp2 + (mdp2.vl + theme(legend.position = "none"))+ plot_layout(widths=c(4,3))
```

```{r}
rse$Group <- mds[colnames(rse),"Group"]
metadata(rse)$anno_colors$Group <- cols
rda$sig <- rda$FDR<0.05
rvol <- ggplot(rda, aes(logFC, -log10(FDR))) + 
   ggrastr::geom_point_rast(size=0.8, colour="darkgrey") + 
   geom_point(data=rda[rda$sig,], size=0.8, aes(colour=sig)) + 
   scale_color_manual(values=TFcols) +
   mytheme() + theme(legend.position = "none") + coord_cartesian(xlim=c(-1.1,1.1)) +
  scale_x_continuous(breaks=c(-1,0,1))
hm <- grid.grabExpr(draw(sechm(rse[,order(rse$Group, rse$Treatment)], row.names(rda)[which(rda$sig)], assayName = "log2FC", gaps_at="TimePoint", top_annotation=c("Group","Sex"), hmcols = c("blue","black","yellow")), merge=FALSE))#, annotation_legend_side="bottom", heatmap_legend_side="bottom"))
```


```{r, fig.width=4, fig.height=5}
p <- promoters(genes(ensdb, filter=GeneNameFilter("Fkbp5")), 2000, 1000)
bws <- split(paste0("../bulkATAC/tracks/",row.names(mds),".bw"), mds$Group)[levels(mds$Group)]
#names(bws) <- gsub("+","+\n",names(bws),fixed=TRUE)

gp <- list(col.line="grey40", col=NULL, fill="#000000", rotation.title=0)
fkbp5 <- grid.grabExpr(plotSignalTracks(bws, p, ensdb, colors=cols, tracks.params = list(ylim=c(0,0.33),transformation=\(x) x, yTicksAt=c(0,0.3), cex.axis=0.55), aggregation="heatmap+mean", genes.params = gp, background.title=NA, cex.title = 0.6), wrap.grobs = TRUE, wrap = TRUE)
fkbp5b <- grid.grabExpr(plotSignalTracks(bws, p, ensdb, colors=cols, tracks.params = list(ylim=c(0,0.34^2),transformation=\(x) x^2, yTicksAt=c(0,0.1), cex.axis=0.55), aggregation="heatmap+mean", genes.params = gp, background.title=NA, cex.title = 0.6), wrap.grobs = TRUE, wrap = TRUE)

p <- promoters(genes(ensdb, filter=GeneNameFilter("Phyhd1")), 2000, 1000)

Phyhd1 <- grid.grabExpr(plotSignalTracks(bws, p, ensdb, colors=cols, genes.params=gp,  tracks.params = list(ylim=c(0,0.2), yTicksAt=c(0,0.1), cex.axis=0.5), aggregation="heatmap+mean", cex.title=0.6), wrap.grobs=TRUE, wrap = TRUE)


# p <- promoters(genes(ensdb, filter=GeneNameFilter("Hif3a")), 2000, 1500)
# plotSignalTracks(bws, p, ensdb, colors=cols, genes.params=gp,  tracks.params = list(ylim=c(0,0.15), yTicksAt=c(0,0.15)), aggregation="heatmap+mean")
```


```{r}
gse <- readRDS("../bulkATAC/gene.top2kbPromoter.SE.rds")
d <- getDEA(gse, "stress")
d$sig <- d$FDR<0.05 & abs(d$logFC)>0.1
d$gene <- row.names(d)
gvol <- ggplot(d, aes(logFC, -log10(FDR), label=gene)) + 
   ggrastr::geom_point_rast(size=0.8, colour="darkgrey") + 
   geom_point(data=d[d$FDR<0.05,], size=0.8, aes(colour=sig)) + 
   ggrepel::geom_text_repel(data=head(d[d$sig,],12), min.segment.length = 0, size=3) + 
   scale_color_manual(values=TFcols) +
   mytheme() + theme(legend.position = "none") + coord_cartesian(xlim=c(-1.1,1.1)) +
  scale_x_continuous(breaks=c(-1,0,1))
```


```{r}
sechm(gse[,order(gse$Group, gse$Treatment)], d$gene[d$sig])
```

```{r}
rna <- readRDS("../rnaDEAs/all.unspliced.SE.rds")
rna2 <- readRDS("../rnaDEAs/all.full.SE.rds")
# rna <- rna[,rna$Treatment!="20days CRS" & rna$TimePoint %in% c("0min","45min")]
# rna$Treatment <- droplevels(rna$Treatment)
# rna$TimePoint <- droplevels(rna$TimePoint)
# dds <- calcNormFactors(DGEList(assay(rna)))
# mm <- model.matrix(~Sex+TimePoint*Treatment, data=as.data.frame(colData(rna)))
# dds <- dds[filterByExpr(dds,mm),]
# dds <- estimateDisp(dds,mm)
# fit <- glmFit(dds,mm)
# str <- as.data.frame(topTags(glmLRT(fit, "TimePoint45min"),Inf))
# using the 10 days only doesn't improve the correlation, so we use all samples to get more sig genes

comDat <- function(str, atac, ghigh=c("Fam107a")){
  sig <- intersect(row.names(str)[str$FDR<0.05 & abs(str$logFC)>log2(1.1)], row.names(d))
  sig2 <- union(sig, intersect(row.names(str),row.names(d)[d$FDR<0.01 & abs(d$logFC)>0.2 & !grepl("^Gm|Rik",row.names(d))]))
  m <- merge(d[sig2,c("logFC","PValue","FDR")], str[sig2,c("logFC","PValue","FDR")], by="row.names", suffix=c(".atac",".rna"))
  m$meanlfdr <- (-log10(m$FDR.atac)-log10(m$FDR.rna))/2
  m$minlfdr <- pmin(-log10(m$FDR.atac),log10(m$FDR.rna))
  m[order(m$minlfdr),]
}

gh <- c("Fam107a", "Cdkn1a", "Sgk1", "Nfkbia", "Mt2", "Htra1", "Hif3a", "Fkbp5","Phyhd1")

d1 <- comDat(getDEA(rna, "45min"), d, ghigh=gh)
d1$RNA <- "Active transcription\nat 45min (unspliced)"
d2 <- comDat(getDEA(rna2, "1h30"), d, ghigh=gh)
d2$RNA <- "Full transcriptome\nat 1h30 (spliced)"
m <- rbind(d1,d2)

ct1 <- cor.test(d1$logFC.atac,d1$logFC.rna)
ct2 <- cor.test(d2$logFC.atac,d2$logFC.rna)
ctlab <- data.frame(RNA=c(d1$RNA[1],d2$RNA[2]),
                    lab=c(paste0("cor=",round(ct1$estimate,2)," (p~",format(ct1$p.value,digits=1),")"),
                          paste0("cor=",round(ct2$estimate,2)," (p~",format(ct2$p.value,digits=1),")")))


rnacor <- ggplot(m, aes(logFC.rna, logFC.atac, label=Row.names)) + 
  geom_hline(yintercept = 0, colour="darkgrey") + geom_vline(xintercept = 0, colour="darkgrey") +
    geom_point(aes(alpha=-log10(FDR.rna), size=-log10(FDR.atac))) + facet_wrap(~RNA, scales="free_x") + 
    mytheme() + ggrepel::geom_text_repel(data=m[m$Row.names %in% gh,], min.segment.length = 0, max.overlaps = 30) +
    geom_text(data=ctlab, aes(label=lab), x=max(m$logFC.rna), y=min(m$logFC.atac), hjust=1, vjust=0) +
  labs(x="RNA log2(foldchange)", y="ATAC promoter log2(foldchange)") + 
  theme(legend.position = "bottom", legend.title.position = "top")
```


```{r}
dev <- readRDS("../bulkATAC/CVdev.SE.rds")
dev <- dev[,order(dev$Group, dev$Treatment, dev$Sex)]
colData(dev)[,c("Treatment","TimePoint","Group")] <- colData(rse)[colnames(dev),c("Treatment","TimePoint","Group")]
dev$Group <- mds[colnames(dev),"Group"]
#levels(dev$Group) <- gsub("+","+\n",levels(dev$Group), fixed=TRUE)
names(cols) <- levels(dev$Group)
d <- getDEA(dev, "stress", homogenize = TRUE)
d$motif <- row.names(d)
movol <- ggplot(d, aes(logFC, -log10(FDR), label=motif)) + 
   geom_point(size=0.8, aes(colour=FDR<0.05)) + 
   ggrepel::geom_text_repel(data=head(d[d$FDR<0.05,],12), min.segment.length = 0) + 
   scale_color_manual(values=TFcols) +
   mytheme() + theme(legend.position = "none") + xlab("Relative difference")

# moeg <- ggplot(meltSE(dev, c("MEF2D","GCR","JUND")), aes(TimePoint, corrected, fill=Group)) + 
#   geom_boxplot() + facet_wrap(~feature, scales="free_y") + mytheme(xax = FALSE) +
#   scale_fill_manual(values=cols) + labs(y="Relative motif accessibility", x=NULL)

moeg <- ggplot(meltSE(dev, c("MEF2D","GCR","JUND")), aes(TimePoint, corrected, fill=Group)) + 
  geom_boxplot() + facet_wrap(~feature, nrow=2, scales="free") + mytheme(xax = FALSE) +
  scale_fill_manual(values=cols) + labs(y="Relative motif accessibility", x=NULL) +
  theme(legend.position = c(1, 0),
        legend.justification = c(1, 0))
```

```{r, fig.width=12, fig.height=14}
pdf("Figure4.pdf", width=12, height=14, bg="white", version="1.4")
#png("Figure4.png", width=12, height=14, units = "in", res = 300)
plot_grid(
  plot_grid(NULL, hm, nrow=1, scale=0.95, labels="AUTO", rel_widths = c(5,4)),
  plot_grid(
    plot_grid(plot_grid(rvol + ggtitle("Regions"), gvol + ggtitle("Promoters"), nrow=1, scale=0.95, labels=c("C","D")),
              plot_grid(rnacor, labels="F", scale=0.95),
              plot_grid(movol, moeg, nrow=1, labels=c("G","H"), scale=0.95, rel_widths = c(3,4)),
              nrow=3, rel_heights = c(4,5,4)),
    plot_grid(fkbp5b,Phyhd1,nrow=2,labels=c("Fkbp5", "Phyhd1"), scale=0.95, label_x = 0.5),
    nrow=1, rel_widths = c(5,4), labels=c(NA,"E")),
  nrow=2, rel_heights = c(1.8,7)
)
dev.off()
```

Just the volcanos:

```{r}
pdf("Figure4_part.pdf", width=12, height=14, bg="white")
#png("Figure4.png", width=12, height=14, units = "in", res = 300)
plot_grid(
  plot_grid(NULL, NULL, nrow=1, scale=0.95, labels="AUTO", rel_widths = c(5,4)),
  plot_grid(
    plot_grid(plot_grid(rvol + ggtitle("Regions"), gvol + ggtitle("Promoters"), nrow=1, scale=0.95, labels=c("C","D")),
              plot_grid(rnacor, labels="F", scale=0.95),
              plot_grid(movol, moeg, nrow=1, labels=c("G","H"), scale=0.95, rel_widths = c(3,4)),
              nrow=3, rel_heights = c(4,5,4)),
    plot_grid(NULL,NULL,nrow=2,labels=c("Fkbp5", "Phyhd1"), scale=0.95, label_x = 0.5),
    nrow=1, rel_widths = c(5,4), labels=c(NA,"E")),
  nrow=2, rel_heights = c(1.8,7)
)
dev.off()

```



```{r,eval=FALSE}
rna1 <- getDEA(rna, c("45min"))
rna2 <- getDEA(rna2, "1h30")
m <- merge(rna1,rna2,by="row.names",suffix=c(".45",".1h30"))
m <- m[which(m$FDR.1h30<0.01 | m$FDR.45<0.05),]
cls <- readRDS("../rnaDEAs/full.dampingClusters.rds")
tmp <- getDEA(gse, "CRS")[m$Row.names[which(m$logFC.1h30>0 & m$FDR.1h30<0.05)],]
hist(cls[[3]]
     
rd  <- rowData(rse)
tmp <- getDEA(rse[rd$nearestTSS.gene_name %in% cls[[3]] & abs(rd$distance2nearestTSS)<10000, ], "interaction")$logFC
tmp2 <- getDEA(rse, "interaction")$logFC
tmp3 <- getDEA(rse[rd$nearestTSS.gene_name %in% row.names(gse) & abs(rd$distance2nearestTSS)<10000, ], "interaction")$logFC
wilcox.test(tmp,tmp2)

flv <- c("cluster1","all peaks","peaks <10kb\nof expressed\n genes")
d <- data.frame(set=rep(factor(flv,flv),c(length(tmp),length(tmp2),length(tmp3))), 
                rank=c(rank(tmp)/length(tmp), rank(tmp2)/length(tmp2), rank(tmp3)/length(tmp3)),
                logFC=c(tmp,tmp2,tmp3))

ggplot(d, aes(logFC, rank, group=set, colour=set)) + geom_line(linewidth=1.15) + theme_bw() + coord_cartesian(xlim=c(-0.8,0.7)) + labs(x="CRS:ARS interacton logFC", y="Cumulative proportion")
```

