---
title: 'Figure 7 : sc'
output: html_document
date: "2024-10-27"
---


```{r}
suppressPackageStartupMessages({
  library(sechm)
  library(ggplot2)
  library(patchwork)
  library(RColorBrewer)
  library(cowplot)
})
theme_set(theme_bw())
source("../functions/functions.R")
cols <- c("Handling +ARS"="darkgrey", "CRS +ARS"="#0000CD96")
```


TFs enriched in neuronal peaks of early-responding genes

```{r}
anno <- readRDS("../scMultiome/neurons.motifMatches.rds")
gff <- rtracklayer::import.gff("/reference/Mus_musculus/cellranger/refdata-cellranger-arc-mm10-2020-A-2.0.0/genes/genes.gtf.gz")
genes <- gff[gff$type=="gene"]
mcols(genes) <- mcols(genes)[,c("gene_name","gene_type")]
di <- distanceToNearest(rowRanges(anno), genes)
rowRanges(anno)$nearest.gene <- genes$gene_name[to(di)]
rowRanges(anno)$nearest.dist <- mcols(di)$distance

rna1 <- readRDS("../scMultiome/GEX_analysis/neuronal.DEA.rds")
expressed <- unique(rna1$gene[!is.na(rna1$logFC)])
rna1 <- rna1[!grepl("Camk2d|Fn1",rna1$cluster_id) & !grepl("Restraint|3h|45min|Any",rna1$Comparison) & rna1$ihw<0.15,]
degs1 <- unique(rna1$gene)
rna2 <- readRDS("../rnaDEAs/all.full.SE.rds")
rna2 <- getDEA(rna2,"45min")
degs2 <- row.names(rna2)[which(rna2$logFC>0.2 & rna2$FDR<0.01)]
degs <- intersect(union(degs1,degs2), expressed)

g <- rowRanges(anno)
w <- which(g$nearest.gene %in% expressed & abs(g$nearest.dist)<10000)
anno <- anno[w]
g <- g[w]
w <- which(g$nearest.gene %in% degs & abs(g$nearest.dist)<10000)
g2 <- g[w]
ga <- anno[w]

sets <- apply(assay(anno),2,FUN=which)
ora <- ORA(w, seq_along(g), sets, excludeUnannotated = FALSE)
ora$expressed <- tools::toTitleCase(tolower(ora$Term)) %in% expressed
ora2 <- ora[ora$expressed,1:6]
p1 <- ggplot(ora2, aes(log2.enrichment, -log10(FDR), label=Term, alpha=FDR<0.05)) + 
  geom_text(size=3) + coord_cartesian(xlim=c(0,2.15)) + 
    scale_alpha_manual(values = c("FALSE"=0.2, "TRUE"=0.75)) + theme_bw() +
  labs(title="Motif enrichment", x="log2(enrichment)", y="-log10(FDR)") +
  theme(legend.position = "bottom")

brainTF <- readRDS("../extDat/brainTF.GR.rds")
btanno <- sapply(split(brainTF, brainTF$TF), FUN=function(x) overlapsAny(g,x))
sets2 <- apply(btanno,2,FUN=which)
btora <- ORA(w, seq_along(g), sets2, excludeUnannotated = FALSE)
p2 <- ggplot(btora, aes(log2.enrichment, -log10(FDR), label=Term, alpha=FDR<0.01)) + 
    geom_point() + ggrepel::geom_text_repel(data=btora[btora$FDR<0.01,], size=3, min.segment.length = 0) + 
    scale_alpha_manual(values = c("FALSE"=0.2, "TRUE"=0.75)) + theme_bw() + 
  labs(title="ChIP binding sites enrichment", x="log2(enrichment)", y="-log10(FDR)") +
  theme(legend.position = "bottom")
btoraB <- ORA(w, seq_along(g), sets2, excludeUnannotated = TRUE)
ora2$Term <- gsub("^GCR$","NR3C1",ora2$Term)
m <- merge(ora2, btora, by="Term", suffix=c(".motif",".chip"))
m$meanFDR <- rowMeans(m[,grep("FDR",colnames(m))])
m$meanLE <- rowMeans(m[,grep("enrichment",colnames(m))])
p3a <- ggplot(m, aes(log2.enrichment.chip, log2.enrichment.motif,
                     alpha=meanFDR, label=Term, fontface=ifelse(meanFDR<0.01,"bold","plain"))) + 
  geom_text() + coord_cartesian(xlim=c(NA, 1.1)) + theme_bw() + 
  scale_alpha_continuous(range=c(1,0.1), breaks=scales::pretty_breaks(n=3)) +
  labs(title="Combined", y="motif log2(enrichment)", x="ChIP log2(enrichment)") +
  theme(legend.position = "bottom")
```


```{r}
btf <- readRDS("../scMultiome/ATAC/brainTF_fastMLM.SE.rds")
#btf2 <- readRDS("../scMultiome/ATAC/brainTF_3TFs_fastMLM.SE.rds")
btf <- updateSEforPlots(btf[,grep("CA1|Cntnap4|Vip", btf$cluster_id)])
#btf2 <- updateSEforPlots(btf2[,grep("CA1|Cntnap4|Vip", btf2$cluster_id)])
d <- meltSE(btf, c("CREB1"))
btfp <- ggplot(d, aes(TimePoint, scaledLFC, fill=Treatment)) + 
  stat_summary(fun=mean, geom="line", linewidth=1.2, aes(x=as.integer(TimePoint), group=Treatment, colour=Treatment))  + 
  geom_boxplot() + facet_grid(~cluster_id) + theme_bw() + labs(y="Inferred TF activity") +
  scale_color_manual(values=cols) + scale_fill_manual(values=cols) + scale_x_discrete() +
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5))

# btf <- readRDS("../scMultiome/ATAC/brainTF_fastMLM.SE.rds")
# btf <- updateSEforPlots(btf[,grep("CA1", btf$cluster_id)])
# d <- meltSE(btf, c("NR3C1"))
# btfp2 <- ggplot(d, aes(TimePoint, scaledLFC, fill=Treatment)) + 
#   stat_summary(fun=mean, geom="line", linewidth=1.2, aes(x=as.integer(TimePoint), group=Treatment, colour=Treatment))  + 
#   geom_boxplot() + facet_grid(~cluster_id) + theme_bw() + labs(y="Inferred TF activity") +
#   scale_color_manual(values=cols) + scale_fill_manual(values=cols) + scale_x_discrete() +
#   theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5))

```




```{r, fig.width=10, fig.height=5}
res <- readRDS("../scMultiome/GEX_analysis/neuronal.DEA.rds")
pb <- readRDS("../scMultiome/GEX_analysis/neuronal.celltype.PB.rds")
pb <- updateSEforPlots(pb)
pb$celltype2 <- gsub("/","\n",pb$celltype)
cls <- as.character(res$cluster_id[which(res$gene=="Fkbp5" & res$ihw<0.05)])
d1 <- meltSE(pb[,pb$celltype %in% cls], "Fkbp5", rowDat.columns = NA)
d1$class <- "neuronal"

resn <- readRDS("../scMultiome/GEX_analysis/nonNeuronal.DEA.rds")
pbn <- readRDS("../scMultiome/GEX_analysis/nonNeuronal.celltype.PB.rds")
pbn$celltype2 <- pbn$celltype
pbn <- log2FC(pbn, "logcpm", pbn$Treatment=="Handling" & pbn$TimePoint=="Control", by=pbn$celltype2)
cls <- as.character(resn$cluster_id[which(resn$gene=="Fkbp5" & resn$ihw<0.01)])
d2 <- meltSE(pbn[,pbn$celltype %in% cls], "Fkbp5", rowDat.columns = NA)
d2$class <- "non-neuronal"
i <- intersect(colnames(d1), colnames(d2))
d <- rbind(d1[,i], d2[,i])
d$cluster_id <- d$celltype
d$celltype2 <- factor(d$celltype2, unique(d$celltype2[order(d$class, d$celltype2)]))

resa <- rbind(res[grep("Any",res$Comparison),],resn[grep("Any",resn$Comparison),])
resa$feature <- resa$gene
m <- merge(d[!duplicated(d[,c("cluster_id","feature")]),], resa[resa$Comparison=="stressAny",])
m$lab <- paste0("ARS q=", format(m$ihw, digit=1),"\n",
                "CRS q=", format(merge(m[,1:2], resa[resa$Comparison=="treatAny",])$ihw, digit=1))

f1 <- ggplot(d, aes(TimePoint, scaledLFC)) + 
  stat_summary(fun=mean, geom="line", linewidth=1.2, aes(x=as.integer(TimePoint), group=Treatment, colour=Treatment))  +
  geom_boxplot(aes(fill=Treatment)) + facet_wrap(~celltype2, nrow=2) + theme_bw() + labs(y="Fkbp5 scaled log2(FC)") +
  geom_text(data=m, y=2.7, x=0.5, aes(label=lab), size=3, colour="black", hjust=0, vjust=1) +
  scale_color_manual(values=cols) + scale_fill_manual(values=cols) + scale_x_discrete() +
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5))
f1
```

```{r}
cls <- as.character(res$cluster_id[which(res$gene=="Egr1" & res$ihw<0.05)])
d <- meltSE(pb[,pb$celltype %in% cls], "Egr1", rowDat.columns = NA)
d$cluster_id <- d$celltype

resa <- res[grep("Any",res$Comparison),]
resa$feature <- resa$gene
m <- merge(d[!duplicated(d[,c("cluster_id","feature")]),], resa[resa$Comparison=="stressAny",])
m$lab <- paste0("ARS q=", format(m$ihw, digit=1),"\n",
                "CRS q=", format(merge(m[,1:2], resa[resa$Comparison=="treatAny",])$ihw, digit=1))

egr1 <- ggplot(d, aes(TimePoint, logcpm)) + 
  stat_summary(fun=mean, geom="line", linewidth=1.2, aes(x=as.integer(TimePoint), group=Treatment, colour=Treatment))  +
  geom_boxplot(aes(fill=Treatment)) + facet_wrap(~celltype2, nrow=1) + theme_bw() + labs(y="Egr1 log(CPM)") +
  geom_text(data=m, y=0, x=4.5, aes(label=lab), size=3, colour="black", hjust=1, vjust=0) +
  scale_color_manual(values=cols) + scale_fill_manual(values=cols) + scale_x_discrete() +
  theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5))

```


# Correlation between GR expression and change in activity

```{r}
pbn <- readRDS("../scMultiome/GEX_analysis/nonNeuronal.broad.PB.rds")
d1 <- meltSE(pb, "Nr3c1", rowDat.columns = NA)
d2 <- meltSE(pbn, "Nr3c1", rowDat.columns = NA)
ag <- aggregate(rbind(d1[,c("logcpm","ncells")], d2[,c("logcpm","ncells")]), by=list(celltype=c(as.character(d1$celltype2), as.character(d2$broadClass))), FUN=median)

load("../scMultiome/ATAC/motifAcc.RData")

se <- readRDS("../scMultiome/ATAC/celltype_cvDevs_neuronal.rds")
dd <- meltSE(se1, "GCR", rowDat.columns = NA)
dd <- dd[dd$TimePoint=="45min",]
ag2 <- aggregate(dd[,c("centered","log2FC","scaledLFC")], by=list(celltype=gsub("/","\n",dd$cluster_id)), FUN=median)
m <- merge(ag,ag2,by="celltype")
m$celltype2 <- gsub("\\..+","",m$celltype)

mod <- summary(lm(log2FC~logcpm, data=m))
otherCols <- c(InhN="#E69F00", astrocytes="#009E73",  microglia="#F0E442", oligodendrocyte="#56B4E9",  OPC="#0072B2", ExN="#D55E00", vascular="#CC79A7")
grcor <- ggplot(m, aes(logcpm, log2FC)) + geom_smooth(method=MASS::rlm, alpha=0.2) + 
  theme_bw() + geom_point(aes(colour=celltype2), size=3) +
  scale_color_manual(values=otherCols) + 
  annotate("text", label=paste0("p=",round(coef(mod)[2,4],3)),
           x=3.5,y=23,hjust=0) +
  labs(x="Nr3c1 (GR) expression (logcpm)", y="Relative change in\nGR motif accessibility",
       colour="Broad cell type")
```



```{r, fig.width=11, fig.height=16}
pdf("Supp_final.pdf", width=11, height=16)
plot_grid(
  plot_grid(btfp + ggtitle("ChIP-based inferred CREB activity"), labels="A", scale=0.95),
  plot_grid(p1, p2, p3a, labels=LETTERS[2:4], scale=0.95, nrow=1),
  plot_grid(egr1 + theme(legend.position = "bottom"), labels="E", scale=0.95),
  plot_grid(f1, labels="F", scale=0.95),
  plot_grid(grcor, NA, labels="G", scale=0.95),
  nrow=5, rel_heights = c(1.3,2,1.5,2.1,1.3)
)
dev.off()
```

